File: src/main/java/org/lucidfox/questfiller/parser/QuestParser.java
package org.lucidfox.questfiller.parser;

import static org.lucidfox.questfiller.parser.ParseUtils.collectTextUntilNextTag;
import static org.lucidfox.questfiller.parser.ParseUtils.getFirstWithOwnText;
import static org.lucidfox.questfiller.parser.ParseUtils.getRegexGroup;
import static org.lucidfox.questfiller.parser.ParseUtils.textOf;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Stream;

import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.nodes.Node;
import org.jsoup.nodes.TextNode;
import org.jsoup.select.Elements;
import org.lucidfox.questfiller.model.core.CharacterClass;
import org.lucidfox.questfiller.model.core.Faction;
import org.lucidfox.questfiller.model.core.ItemReward;
import org.lucidfox.questfiller.model.core.Race;
import org.lucidfox.questfiller.model.quest.Quest;

final class QuestParser implements IParser<Quest> {
	// Quest categories for which most quests scale with level
	private static final Set<String> LEGION_SCALING_QUEST_CATEGORIES = new HashSet<>(Arrays.asList(
			"Azsuna", "Val'sharah", "Highmountain", "Stormheim", "Artifact"
	));
	
	private final ParserContext context;
	
	QuestParser(final ParserContext context) {
		this.context = context; 
	}
	
	public Quest parse(final Document html) {
		final Quest quest = new Quest();
		
		final String url = html.select("link[rel=canonical]").attr("href");
		final String idStr = getRegexGroup(url, "/quest=([0-9]+)/", 1).get();
		quest.setId(Integer.parseInt(idStr));
		
		final Element mainContainer = html.select("#main-contents div.text").first();
		final Element questName = html.select("h1.heading-size-1").first();
		quest.setName(questName.text());
		
		parseCategory(quest, html);
		parseObjectives(quest, mainContainer, questName);
		parseQuestText(quest, mainContainer, html);
		parseMoney(quest, mainContainer);
		parseRewards(quest, mainContainer);
		parseGains(quest, mainContainer);
		parseInfobox(quest, html);
		parseSeries(quest, html);
		return quest;
	}

	private void parseCategory(final Quest quest, final Document html) {
		// Category taken from breadcrumb, which Wowhead draws with JS :(
		final int[] categoryIds = ParseUtils.getCategoryIds(html);
		// last number in list is the category id
		final int categoryId = categoryIds[categoryIds.length - 1];
		quest.setCategory(context.getQuestCategory(categoryId));
	}

	private void parseObjectives(final Quest quest, final Element mainContainer, final Element questName) {
		// Objectives section
		Node objectivesNode = questName.nextSibling();
		
		// Objectives text is the first non-empty text node immediately following the header
		while (!(objectivesNode instanceof TextNode) || ((TextNode) objectivesNode).text().trim().isEmpty()) {
			objectivesNode = objectivesNode.nextSibling();
		}
		
		final Node beforeObjectives = objectivesNode.previousSibling();
		
		// If there is a h2.heading-size-3 right before the "objectives" text, it is probably not objectives,
		// but rather progress or completion, like on the quest "Draenei Tail"
		if (!(beforeObjectives instanceof Element && ((Element) beforeObjectives).tagName().equals("h2")
				&& ((Element) beforeObjectives).hasClass("heading-size-3"))) {
			quest.setObjectives(((TextNode) objectivesNode).text().trim());
		}
		
		// Objective completion stages
		final Elements iconlists = mainContainer.select("table.iconlist");
		final Element stagesTable = iconlists.first();
		
		if (stagesTable != null) {
			// Remove any subtables
			stagesTable.select("table.iconlist").remove();
			
			for (final Element stageLink : stagesTable.getElementsByTag("a")) {
				// Find the innermost td element enclosing the a, and add its whole text
				Element parent = stageLink.parent();
				
				while (!parent.tagName().equals("td")) {
					parent = parent.parent();
				}
				
				quest.getStages().add(parent.text());
			}

			// Suggested players
			final Element suggestedPlayers = stagesTable.getElementsContainingOwnText("Suggested players:").first();
			
			if (suggestedPlayers != null) {
				String playerCountStr =
						getRegexGroup(suggestedPlayers.ownText(), "Suggested players: ([0-9]+)", 1).get();
				quest.setGroupSize(Integer.parseInt(playerCountStr));
			}
		}
		
		// Provided items
		if (iconlists.size() >= 2) {
			final Element maybeProvided = iconlists.get(1);
			final Node before = maybeProvided.previousSibling();
			
			if (before instanceof TextNode && ((TextNode) before).text().contains("Provided")) {
				maybeProvided.select("table.iconlist").remove();
				
				for (final Element itemLink : maybeProvided.getElementsByTag("a")) {
					quest.getProvidedItems().add(itemLink.text());
				}
			}
		}
	}

	private void parseQuestText(final Quest quest, final Element mainContainer, final Document html) {
		// Description section
		final Elements headingsSize3 = mainContainer.select("h2.heading-size-3");
		
		getFirstWithOwnText(headingsSize3, "Description").ifPresent(descriptionHeading -> {
			quest.setDescription(collectTextUntilNextTag(descriptionHeading, "h2").trim());
		});
		
		// Progress section
		final Element progressHeading = html.getElementById("lknlksndgg-progress");
		
		if (progressHeading != null) {
			quest.setProgress(textOf(progressHeading));
		} else {
			getFirstWithOwnText(headingsSize3, "Progress").ifPresent(fallbackProgressHeading -> {
				quest.setProgress(collectTextUntilNextTag(fallbackProgressHeading, "h2").trim());
			});
		}
		
		// Completion section
		final Element completionHeading = html.getElementById("lknlksndgg-completion");
		
		if (completionHeading != null) {
			quest.setCompletion(textOf(completionHeading));
		} else {
			getFirstWithOwnText(headingsSize3, "Completion").ifPresent(fallbackCompletionHeading -> {
				quest.setCompletion(collectTextUntilNextTag(fallbackCompletionHeading, "h2").trim());
			});
		}
	}
		
	private void parseMoney(final Quest quest, final Element mainContainer) {
		// Money rewards
		final List<Element> allMoneyElements = new ArrayList<Element>();
		allMoneyElements.addAll(mainContainer.select("span.moneygold"));
		allMoneyElements.addAll(mainContainer.select("span.moneysilver"));
		allMoneyElements.addAll(mainContainer.select("span.moneycopper"));
		
		// Find the leftmost money element (so we parse only the first group - second group is max level rewards)
		if (!allMoneyElements.isEmpty()) {
			int money = 0;
			Element leftEl = null;
			
			for (final Element el: allMoneyElements) {
				if (leftEl == null || el.siblingIndex() < leftEl.siblingIndex()) {
					leftEl = el;
				}
			}
			
			// Parse all directly adjacent span nodes, possibly with whitespace in between
			for (Node node = leftEl;
					(node instanceof Element && ((Element) node).tagName().equals("span"))
						|| (node instanceof TextNode && ((TextNode) node).text().trim().isEmpty());
					node = node.nextSibling()) {
				if (!(node instanceof Element)) {
					continue;
				}
				
				final Element el = (Element) node;
				
				if (el.hasClass("moneygold")) {
					money += Integer.parseInt(el.ownText()) * 10000;
				} else if (el.hasClass("moneysilver")) {
					money += Integer.parseInt(el.ownText()) * 100;
				} else if (el.hasClass("moneycopper")) {
					money += Integer.parseInt(el.ownText());
				}
			}
			
			quest.setMoney(money);
		}
	}
		
	private void parseRewards(final Quest quest, final Element mainContainer) {
		// Non-money rewards
		final Elements icontabs = mainContainer.select("table.icontab.icontab-box");
		
		for (final Element icontab : icontabs) {
			final Node prevNode = icontab.previousSibling();
			
			if (icontab.id().equals("dynamic-rewards")) {
				collectItemRewards(icontab, quest.getChoiceRewards());
			} else if (prevNode instanceof TextNode) {
				final String prevText = ((TextNode) prevNode).text();
				
				if (prevText.contains("You will receive:") || prevText.contains("You will also receive:")) {
					collectItemRewards(icontab, quest.getNonChoiceRewards());
				} else if (prevText.contains("You will be able to choose one of these rewards:")) {
					collectItemRewards(icontab, quest.getChoiceRewards());
				} else if (prevText.contains("You will learn:")) {
					collectNonItemRewards(icontab, quest.getAbilityRewards());
				} else if (prevText.contains("The following spell will be cast on you:")) {
					collectNonItemRewards(icontab, quest.getBuffRewards());
				} else if (prevText.trim().isEmpty() && isMoneyRewardSpan(icontab.previousElementSibling())) {
					// This is probably an item tucked at the end after money rewards
					collectItemRewards(icontab, quest.getNonChoiceRewards());
				}
			}
		}
	}
	
	// Convenience collection adapter around ParseUtils.collectItemRewards, which expects a Consumer
	private void collectItemRewards(final Element icontab, final Collection<ItemReward> collector) {
		ParseUtils.collectItemRewards(icontab, collector::add);
	}
	
	private void collectNonItemRewards(final Element icontab, final Collection<String> collector) {
		for (final Element link : icontab.getElementsByTag("a")) {
			collector.add(link.ownText());
		}
	}
	
	private boolean isMoneyRewardSpan(final Element element) {
		return "span".equals(element.tagName()) && element.className().matches(".*money(?:gold|silver|copper).*");
	}

	private void parseGains(final Quest quest, final Element mainContainer) {
		// Gains section
		final Elements headingsSize3 = mainContainer.select("h2.heading-size-3");
		final Optional<Element> gains = headingsSize3.stream()
				.filter(el -> el.text().equals("Gains")).findFirst();
		
		if (gains.isPresent()) {
			Node gainsData = gains.get().nextSibling();
			
			while (!(gainsData instanceof Element && ((Element) gainsData).tagName().equals("ul"))) {
				gainsData = gainsData.nextSibling();
			}
			
			Elements divs = ((Element) gainsData).getElementsByTag("div");
			int firstNonXPDiv;
			
			if (divs.first().ownText().contains("experience")) {
				firstNonXPDiv = 1;
				final String xpValue = getRegexGroup(divs.first().ownText(), "([0-9,]*) experience", 1).get();
				quest.setExperience(Integer.parseInt(xpValue.replace(",", "")));
			} else {
				firstNonXPDiv = 0;
			}
			
			for (int i = firstNonXPDiv; i < divs.size(); i++) {
				final Element div = divs.get(i);
				
				if (div.ownText().contains("reputation with")) {
					final String repValue = div.getElementsByTag("span").first().ownText();
					final String faction = div.getElementsByTag("a").first().ownText();
					quest.getReputationGains().put(faction, Integer.parseInt(repValue.replace(",", "")));
				} else {
					// Non-reputation gain
					quest.getOtherGains().add(div.text());
				}
			}
		}
	}
	
	private void parseInfobox(final Quest quest, final Document html) {
		// Infobox section
		final List<String> infoboxLines = ParseUtils.getInfoboxLines(html);
		
		// Pattern-match each infobox line
		for (final String infoboxLine : infoboxLines) {
			getRegexGroup(infoboxLine, "Level: ([0-9]+)", 1).ifPresent(levelStr -> {
				quest.setLevel(Integer.parseInt(levelStr));
			});
			
			getRegexGroup(infoboxLine, "Requires level ([0-9]+)", 1).ifPresent(levelStr -> {
				quest.setLevelRequired(Integer.parseInt(levelStr));
			});
			
			getRegexGroup(infoboxLine, "Type: (.+)", 1).ifPresent(type -> {
				if (type.equals("Artifact")) {
					// This is what wowpedia uses
					quest.setType("Legendary");
				} else {
					quest.setType(type);
				}
			});
			
			getRegexGroup(infoboxLine, "Side: (.+)", 1).ifPresent(side -> {
				switch (side) {
				case "Alliance":
					quest.setFaction(Faction.ALLIANCE);
					break;
				case "Horde":
					quest.setFaction(Faction.HORDE);
					break;
				case "Both":
					quest.setFaction(Faction.NEUTRAL);
					break;
				default:
					System.err.printf("Unknown side %s\n", side);
					// Unknown faction
					quest.setFaction(null);
					break;
				}
			});
			
			getRegexGroup(infoboxLine, "Race: ([0-9]+)", 1).ifPresent(raceId -> {
				quest.setRace(Race.getById(Integer.parseInt(raceId)));
			});
			
			getRegexGroup(infoboxLine, "Class: ([0-9]+)", 1).ifPresent(classId -> {
				quest.setCharacterClass(CharacterClass.getById(Integer.parseInt(classId)));
			});
			
			getRegexGroup(infoboxLine, "Start: (.+)", 1).ifPresent(startEntity -> {
				quest.setStartEntity(startEntity);
			});
			
			getRegexGroup(infoboxLine, "End: (.+)", 1).ifPresent(finishEntity -> {
				quest.setFinishEntity(finishEntity);
			});
			
			getRegexGroup(infoboxLine, "Added in patch ([0-9]+.[0-9]+.[0-9]+)", 1).ifPresent(patch -> {
				quest.setPatchAdded(PatchVersions.getCanonicalVersion(patch));
			});
			
			if ("Repeatable".equals(infoboxLine)) {
				quest.setRepeatable(true);
			} else if ("Sharable".equals(infoboxLine)) {
				quest.setShareable(true);
			} else if ("Not sharable".equals(infoboxLine)) {
				quest.setShareable(false);
			}
			
			// If we still don't know the level, and it's a Legion quest, its level probably scales
			if (quest.getLevel() == null && LEGION_SCALING_QUEST_CATEGORIES.contains(quest.getCategory())) {
				quest.setLevel(100);
			}
		}
	}
	
	private void parseSeries(final Quest quest, final Document html) {
		// Try to determine previous and next quests
		final Element seriesTable = html.select("#sidebar table.series").first();
		
		if (seriesTable == null) {
			return;
		}
		
		// Find the table cell with the current quest, in bold
		final Element ourQuestItem = seriesTable.getElementsByTag("b").first();
		Element ourQuestCell = ourQuestItem.parent();
		
		while (!ourQuestCell.tagName().equals("td")) {
			ourQuestCell = ourQuestCell.parent();
		}
		
		// Find where in the table it is
		final Elements questCells = seriesTable.getElementsByTag("td");
		final int ourQuestIndex = questCells.indexOf(ourQuestCell);
		
		if (ourQuestIndex > 0) {
			final String[] previousQuests = textOf(questCells.get(ourQuestIndex - 1)).split("\n");
			Stream.of(previousQuests).forEach(quest.getPreviousQuests()::add);
		}
		
		if (ourQuestIndex < questCells.size() - 1) {
			final String[] nextQuests = textOf(questCells.get(ourQuestIndex + 1)).split("\n");
			Stream.of(nextQuests).forEach(quest.getNextQuests()::add);
		}
	}
}

File: src/parsers/QuestParser.js
import { $, $$ } from "../common/shortcuts";
import u from "./utils";
import substitutions from "./substitutions";

import questEJS from "../templates/quest.ejs";

import Quest from "../model/quest/Quest";
import ItemReward from "../model/core/ItemReward";
import ReputationGain from "../model/quest/ReputationGain";

const LEGION_SCALING_QUEST_CATEGORIES = new Set([
	"Azsuna", "Val'sharah", "Highmountain", "Stormheim", "Artifact"
]);

/**
 * @return {Quest}
 */
export default function QuestParser(context) {
	this.name = "Quest";
	this.templatePrefix = "q";
	this.template = questEJS;

	/**
	 * @return {Quest}
	 */
	this.parse = function() {
		let quest = new Quest();
		
		let url = $("link[rel=canonical]").getAttribute("href");
		let idStr = u.getRegexGroup(url, "/quest=([0-9]+)/", 1);
		quest.id = parseInt(idStr, 10);
		
		let mainContainer = $("#main-contents > div.text");
		let questName = $("h1.heading-size-1");
		quest.name = questName.textContent;
		
		parseCategory(quest);
		parseObjectives(quest, mainContainer, questName);
		parseQuestText(quest, mainContainer);
		parseMoney(quest, mainContainer);
		parseRewards(quest, mainContainer);
		parseGains(quest, mainContainer);
		parseInfobox(quest);
		parseSeries(quest, mainContainer);
		parseRemoved(quest, mainContainer);

		return quest;
	}

	/**
	 * @param {Quest} quest
	 */
	function parseCategory(quest) {
		let categoryIds = u.getCategoryIds();
		let categoryId = categoryIds[categoryIds.length - 1];
		quest.category = context.questCategories[categoryId.toString()];
	}

	/**
	 * @param {Quest} quest 
	 * @param {Element} mainContainer
	 * @param {Element} questName 
	 */
	function parseObjectives(quest, mainContainer, questName) {
		// Objectives section
		let objectivesNode = questName.nextSibling;

		if (objectivesNode == null) {
			return;
		}
		
		// Objectives text is the first non-empty text node immediately following the header
		while (!(objectivesNode instanceof Text) || objectivesNode.data.trim().length === 0) {
			objectivesNode = objectivesNode.nextSibling;
		}
		
		let beforeObjectives = objectivesNode.previousSibling;
		
		// If there is a h2.heading-size-3 right before the "objectives" text, it is probably not objectives,
		// but rather progress or completion, like on the quest "Draenei Tail"
		if (!(beforeObjectives instanceof Element && u.tagName(beforeObjectives) === "h2"
				&& beforeObjectives.classList.contains("heading-size-3"))) {
			quest.objectives = u.normalize(objectivesNode.data);
		}
		
		// Objective completion stages
		let iconlists = mainContainer.querySelectorAll("table.iconlist");
		
		if (iconlists.length > 0) {
			let stagesTable = iconlists[0];

			// Remove any subtables
			for (let element of stagesTable.querySelectorAll("table.iconlist")) {
				element.parentElement.removeChild(element);
			}

			let tds = stagesTable.getElementsByTagName("td");

			if (!quest.objectives && tds.length > 0 && tds[0].getElementsByTagName("a").length === 0) {
				// No objectives, but first element is not a link. Assume it's the objective.
				quest.objectives = u.normalize(u.textOf(tds[0]));
			}
			
			for (let stageLink of stagesTable.querySelectorAll("td a")) {
				// Find the innermost td element enclosing the a, and add its whole text
				let parent = stageLink.parentElement;
				
				while (u.tagName(parent) !== "td") {
					parent = parent.parentElement;
				}
				
				quest.stages.push(u.normalize(parent.textContent));
			}

			// Suggested players
			let suggestedPlayers = u.getElementContainingOwnText(stagesTable, "td", "Suggested players:");
			
			if (suggestedPlayers != null) {
				let playerCountStr =
						u.getRegexGroup(suggestedPlayers.textContent, "Suggested players: ([0-9]+)", 1);
				quest.groupSize = parseInt(playerCountStr);
			}
		}
		
		// Provided items
		if (iconlists.length >= 2) {
			let maybeProvided = iconlists[1];
			let before = maybeProvided.previousSibling;
			
			if (before instanceof Text && before.textContent.includes("Provided")) {
				for (let element of maybeProvided.querySelectorAll("table.iconlist")) {
					element.parentElement.removeChild(element);
				}
				
				for (let itemLink of maybeProvided.querySelectorAll("td a")) {
					quest.providedItems.push(u.normalize(itemLink.textContent));
				}
			}
		}
	}

	/**
	 * @param {Quest} quest 
	 * @param {Element} mainContainer
	 */
	function parseQuestText(quest, mainContainer) {
		// Description section
		let headingsSize3 = mainContainer.querySelectorAll("h2.heading-size-3");

		if (headingsSize3 == null) {
			return;
		}
		
		u.getFirstWithOwnText(headingsSize3, "Description", descriptionHeading => {
			quest.description = u.normalize(u.collectTextUntilNextTag(descriptionHeading, "h2"));
		});
		
		// Progress section
		let progressHeading = document.getElementById("lknlksndgg-progress");
		
		if (progressHeading != null) {
			quest.progress = u.textOf(progressHeading);
		} else {
			u.getFirstWithOwnText(headingsSize3, "Progress", fallbackProgressHeading => {
				quest.progress = u.normalize(u.collectTextUntilNextTag(fallbackProgressHeading, "h2"));
			});
		}
		
		// Completion section
		let completionHeading = document.getElementById("lknlksndgg-completion");
		
		if (completionHeading != null) {
			quest.completion = u.textOf(completionHeading);
		} else {
			u.getFirstWithOwnText(headingsSize3, "Completion", fallbackCompletionHeading => {
				quest.completion = u.normalize(u.collectTextUntilNextTag(fallbackCompletionHeading, "h2"));
			});
		}
	}
		
	/**
	 * @param {Quest} quest 
	 * @param {Element} mainContainer
	 */
	function parseMoney(quest, mainContainer) {
		quest.money = u.getMoney(mainContainer);
	}
	
	/**
	 * @param {Quest} quest 
	 * @param {Element} mainContainer
	 */
	function parseRewards(quest, mainContainer) {
		// Non-money rewards
		let icontabs = mainContainer.querySelectorAll("table.icontab.icontab-box");
		
		for (let icontab of icontabs) {
			let prevNode = icontab.previousSibling;

			if (prevNode.id === "quest-reward-coin") {
				prevNode = prevNode.previousSibling;
			}
			
			if (icontab.id === "dynamic-rewards" || icontab.id === "choicerewards") {
				collectItemRewards(icontab, quest.choiceRewards);
			} else if (prevNode instanceof Text) {
				let prevText = prevNode.textContent;
				
				if (prevText.includes("You will receive:") || prevText.includes("You will also receive:")) {
					collectItemRewards(icontab, quest.nonChoiceRewards);
				} else if (prevText.includes("You will be able to choose one of these rewards:")) {
					collectItemRewards(icontab, quest.choiceRewards);
				} else if (prevText.includes("You will learn:")) {
					collectNonItemRewards(icontab, quest.abilityRewards);
				} else if (prevText.includes("The following spell will be cast on you:")) {
					collectNonItemRewards(icontab, quest.buffRewards);
				} else if ((prevText.trim().length === 0 || prevText.includes("if completed at level"))
						&& isMoneyRewardSpan(icontab.previousElementSibling)) {
					// This is probably an item tucked at the end after money rewards
					collectItemRewards(icontab, quest.nonChoiceRewards);
				}
			}
		}
	}
	
	// Convenience collection adapter around ParseUtils.collectItemRewards, which expects a Consumer
	/**
	 * @param {Element} icontab
	 * @param {Array.<ItemReward>} collector
	 */
	function collectItemRewards(icontab, collector) {
		u.collectItemRewards(icontab, (item, quantity) => {
			collector.push(new ItemReward(item, quantity, collector.length + 1));
		});
	}
	
	/**
	 * @param {Element} icontab
	 * @param {Array.<string>} collector
	 */
	function collectNonItemRewards(icontab, collector) {
		for (let link of icontab.querySelectorAll("td > a")) {
			collector.push(u.normalize(link.textContent));
		}
	}
	
	/**
	 * @param {Element} element
	 * @return {boolean}
	 */
	function isMoneyRewardSpan(element) {
		return u.tagName(element) === "span" && /money(?:gold|silver|copper)/.test(element.className);
	}

	/**
	 * @param {Quest} quest 
	 * @param {Element} mainContainer
	 */
	function parseGains(quest, mainContainer) {
		// Gains section
		let gainsData = null;

		for (let heading of mainContainer.querySelectorAll("h2.heading-size-3")) {
			if (heading.textContent === "Gains") {
				gainsData = u.findNextElementSibling(heading, el => u.tagName(el) === "ul");
				break;
			}
		}

		if (gainsData) {
			let divs = gainsData.getElementsByTagName("div");
			let firstNonXPDiv;
			
			if (divs[0].textContent.includes("experience")) {
				firstNonXPDiv = 1;
				let xpValue = u.getRegexGroup(divs[0].textContent, "([0-9, ]*) experience", 1);
				quest.experience = parseInt(xpValue.replace(/[ ,]/g, ""));
			} else {
				firstNonXPDiv = 0;
			}
			
			for (let i = firstNonXPDiv; i < divs.length; i++) {
				let div = divs[i];
				
				if (div.textContent.includes("reputation with")) {
					let repValue = div.getElementsByTagName("span")[0].textContent;
					let faction = div.getElementsByTagName("a")[0].textContent;
					let canonicalName = substitutions.getCanonicalReputationFaction(faction);
					quest.reputationGains.push(new ReputationGain(faction,
							canonicalName === faction ? null : canonicalName,
							parseInt(repValue.replace(",", ""))));
				} else {
					// Non-reputation gain
					quest.otherGains.push(u.normalize(div.textContent));
				}
			}
		}
	}
	
	/**
	 * @param {Quest} quest
	 */
	function parseInfobox(quest) {
		// Infobox section
		// Pattern-match each infobox line
		for (let infoboxLine of u.getInfoboxLines(true)) {
			u.getRegexGroup(infoboxLine, "Level: ([0-9]+)", 1, levelStr => {
				quest.level = parseInt(levelStr);
			});
			
			u.getRegexGroup(infoboxLine, "Requires level ([0-9]+)", 1, levelStr => {
				quest.levelRequired = parseInt(levelStr);
			});
			
			u.getRegexGroup(infoboxLine, "Type: (.+)", 1, type => {
				if (type === "Artifact") {
					// This is what wowpedia uses
					quest.type = "Legendary";
				} else {
					quest.type = type;
				}
			});
			
			u.getRegexGroup(infoboxLine, "Side: (.+)", 1, side => {
				if (side === "Both") {
					quest.faction = "Neutral";
				} else {
					quest.faction = side;
				}
			});
			
			u.getRegexGroup(infoboxLine, "Race: ([0-9]+)", 1, raceId => {
				quest.race = context.races[raceId];
			});
			
			u.getRegexGroup(infoboxLine, "Class: ([0-9]+)", 1, classId => {
				quest.characterClass = context.classes[classId];
			});
			
			u.getRegexGroup(infoboxLine, "Start: (.+)", 1, startEntity => {
				quest.startEntity = startEntity;
			});
			
			u.getRegexGroup(infoboxLine, "End: (.+)", 1, finishEntity => {
				quest.finishEntity = finishEntity;
			});
			
			u.getRegexGroup(infoboxLine, "Added in patch ([0-9]+.[0-9]+.[0-9]+)", 1, patch => {
				quest.patchAdded = substitutions.getCanonicalPatchVersion(patch);
			});
			
			if (infoboxLine === "Repeatable") {
				quest.repeatable = true;
			} else if (infoboxLine === "Sharable") {
				quest.shareable = true;
			} else if (infoboxLine === "Not sharable") {
				quest.shareable = false;
			}
			
			// If we still don't know the level, and it's a Legion quest, its level probably scales
			if (!quest.level && LEGION_SCALING_QUEST_CATEGORIES.has(quest.category)) {
				quest.level = "100 - 110";
			}
		}
	}
	
	/**
	 * @param {Quest} quest 
	 * @param {Element} mainContainer
	 */
	function parseSeries(quest, mainContainer) {
		// Try to determine previous and next quests
		let seriesHeader = $("#infobox-series");

		if (!seriesHeader) {
			return;
		}

		let seriesTable = seriesHeader.parentElement.parentElement.querySelector("table.series");

		if (!seriesTable) {
			return;
		}
		
		// Find the table cell with the current quest, in bold
		let ourQuestItem = seriesTable.getElementsByTagName("b")[0];
		let ourQuestCell = ourQuestItem.parentElement;
		
		while (u.tagName(ourQuestCell) !== "td") {
			ourQuestCell = ourQuestCell.parentElement;
		}
		
		// Find where in the table it is
		let questCells = Array.from(seriesTable.getElementsByTagName("td"));
		let ourQuestIndex = questCells.indexOf(ourQuestCell);
		
		if (ourQuestIndex > 0) {
			let previousQuests = u.textOf(questCells[ourQuestIndex - 1]).split("\n");
			previousQuests.forEach(q => quest.previousQuests.push(q));
		}
		
		if (ourQuestIndex < questCells.length - 1) {
			let nextQuests = u.textOf(questCells[ourQuestIndex + 1]).split("\n");
			nextQuests.forEach(q => quest.nextQuests.push(q));
		}
	}
	
	/**
	 * @param {Quest} quest 
	 * @param {Element} mainContainer
	 */
	function parseRemoved(quest, mainContainer) {
		// Set a removed flag if there is an obsolete warning on the quest page
		for (let el of mainContainer.querySelectorAll("b[style=\"color: red\"]")) {
			if (el.textContent.startsWith("This quest was marked obsolete")) {
				quest.removed = true;
			}
		}
	}
}
